pennylane/interfaces/torch.py

```diff
@@ -184,7 +184,21 @@ def backward(ctx_, ddy):  # pragma: no cover
     def backward(ctx, dy):  # pragma: no cover
         """Implements the backwards pass QNode vector-Jacobian product"""
         ctx.dy = dy
-        vjp = dy.view(1, -1) @ ctx.jacobian.apply(ctx, *ctx.saved_tensors)
+
+        dyv = dy.view(1, -1)
+        jac_res = ctx.jacobian.apply(ctx, *ctx.saved_tensors)
+
+        # When using CUDA, dyv seems to remain on the GPU, while the result
+        # of jac_res is returned on CPU, even though the saved_tensors arguments are
+        # themselves on the GPU. Check whether this has happened, and move things
+        # back to the GPU if required.
+        if dyv.is_cuda or jac_res.is_cuda:
+            if not dyv.is_cuda:
+                dyv = torch.as_tensor(dyv, device=jac_res.get_device())
+            if not jac_res.is_cuda:
+                jac_res = torch.as_tensor(jac_res, device=dyv.get_device())

```

Comment:
It occurs to me that we could factor this out as a utility function, so that this logic can be re-used where needed in the future:

```python
def match_gpu_device(tensors):
    """Ensure that tensors are placed on the same device.

    This function checks whether any tensor within the input list of
    tensors is on a GPU. If this is the case, all tensors are moved
    to the location of the *first CUDA tensor in the list*.

    If no tensors are on a GPU, the tensor list is returned unchanged.

    Args:
        tensors (sequence[torch.Tensor]): list of input tensors
    """
    device = None

    for t in tensors:
        if t.is_cuda:
            device = t.get_device()
            break

    if device is None:
        return

    for t in tensors:
        if t.get_device() is not device:
            torch.as_tensor(t, device=device)
```